lib/spack/spack/test/build_environment.py

```diff
@@ -337,9 +337,9 @@ def test_wrapper_variables(
         ]
     )
     cuda_include_dirs = cuda_headers.directories
-    assert posixpath.join("prefix", "include") in cuda_include_dirs
+    assert os.path.join("prefix", "include") in cuda_include_dirs
     assert (
-        posixpath.join("prefix", "include", "cuda", "std", "detail", "libcxx", "include")
+        os.path.join("prefix", "include", "cuda", "std", "detail", "libcxx", "include")
```

Comment:
I'm not quite sure, but this seems like it would break on windows. Previously, I believe `os.path.join()` would join each argument with `\` on windows, but here it would use a `/`. Is this intentional?

Reply:
This is the same change as above where prior to now, `HeaderList`'s behavior was to pipe through whatever style path was provided and now it marshals paths to their appropriate system style.

Reply:
That sounds like a bit of an unintended API break, even though it sounds usually beneficial.  To make sure I understand, is the behavior now:

1. use the appropriate system path type by default, but switch to the one provided if the caller explicitly provides one, or
2. always use the system path type?

I ask because, if we eventually support cross-compilation or even just want to support working with paths in other conventions the latter might be an issue.

Reply:
What is the intended behavior of the `HeaderList` class' directories attribute? 
For some context, prior to the Windows support changes, paths were all unix style simply because that's what was being used. Windows support introduced path marshalling to use system path styles, but only introduced support for that marshalling in the setter for `directories`, so if that was never invoked, path style in was path style out simply as a side effect of an incomplete refactor, the intent was to make all paths system style dependent. This was an oversight on my part, and introduced some very obfuscated inconsistency in behavior that has no conceptual demarcation. I should have also handled the path marshalling in `HeaderList._default_directories` as well, which this PR corrects, and which is where this change comes in. 

From my perspective it doesn't seem there was an established convention to break from here, and if the convention was to give the user a requested path type, that functionality does not currently exist in the API, and in fact the user may not get the path it passed in in some contexts, which is functionality the caller is currently blind to. In general, IMO, Spack should keep internal representations of paths to be `pathlib` objects until those objects need to be passed out of Spack, either by a filesystem API call or an env variable, or some other avenue. At that point, the `pathlib` object should be cast to the appropriate style/usage and processed as needed by the calling context. This makes the "`HeaderList` should return posix or Windows style paths a moot point, as a caller can simply run `.as_posix()` or `as_windows()` as needed. I actually think `HeaderList` should rely on `PurePath` not `Path` as it's representing abstract paths not direct filesystem operations, which callers should cast to the appropriate concrete path style. 
Unfortunately while the `pathlib` refactor is still piecemeal we have to cast to string when returning things from `HeaderList`, which requires us to pick a lane for the time being. The current behavior is to marshal all paths to their system style, and IMO it should stay this way, as we should in general expect a consistent style of path representation internal to Spack, and alter that on a case by case basis in the calling context when making calls to an endpoint outside of Spack. The alternative is adding a option to the HeaderList API to request unix or Windows style paths, which IMO adds potential points of fragility/confusion and a lot more logic inside the `HeaderList` class.